src/transformers/models/afmoe/modular_afmoe.py

```diff
@@ -0,0 +1,640 @@
+    """
+    Token-choice top-K router for MoE routing.
+
+    This router assigns each token to the top-K experts based on learned routing scores.
+    It supports both sigmoid and softmax scoring functions.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.top_k = config.num_experts_per_tok
+        self.num_experts = config.num_experts
+        self.score_func = config.score_func
+        self.route_norm = config.route_norm
+        self.route_scale = config.route_scale
+        self.gate = nn.Linear(config.hidden_size, config.num_experts, bias=False)
+
+    def forward(self, hidden_states, expert_bias: torch.Tensor | None):
+        _, _, hidden_dim = hidden_states.shape
+        hidden_states = hidden_states.view(-1, hidden_dim)
+
+        scores = self.gate(hidden_states)
+
+        # Apply scoring function in float32 for stability
+        if self.score_func == "sigmoid":
+            scores = torch.sigmoid(scores.to(torch.float32))
+        else:
+            scores = F.softmax(scores.to(torch.float32), dim=-1)
+
+        if expert_bias is not None:
+            _, selected_experts = torch.topk(scores + expert_bias, k=self.top_k, dim=1)
+            top_scores = scores.gather(dim=1, index=selected_experts)
+        else:
+            top_scores, selected_experts = torch.topk(scores, k=self.top_k, dim=1)
+
+        # Normalize weights if using sigmoid
+        if self.score_func == "sigmoid" and self.route_norm:
+            denominator = top_scores.sum(dim=-1, keepdim=True) + 1e-20
+            top_scores = top_scores / denominator
+
+        top_scores = top_scores * self.route_scale
+        return top_scores, selected_experts
+
+
+class AfmoeMoE(nn.Module):
+    """
+    Mixture of Experts (MoE) module for AFMoE.
+
+    This module implements a sparse MoE layer with both shared experts (always active) and
+    routed experts (activated based on token-choice routing).
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.router = AfmoeTokenChoiceRouter(config)
+
+        self.shared_experts = None
+        if config.num_shared_experts > 0:
+            self.shared_experts = AfmoeMLP(config, config.moe_intermediate_size * config.num_shared_experts)
+        self.experts = nn.ModuleList(
+            [AfmoeMLP(config, intermediate_size=config.moe_intermediate_size) for _ in range(config.num_experts)]
+        )
+        self.expert_bias = nn.Parameter(torch.zeros(config.num_experts, dtype=torch.float32), requires_grad=False)
+
+    def forward(self, hidden_states):
+        batch_size, seq_len, hidden_dim = hidden_states.shape
+        hidden_states_flat = hidden_states.view(-1, hidden_dim)
+
+        # Get routing decisions
+        top_scores, selected_experts = self.router(hidden_states, self.expert_bias)
+
+        # Process through shared experts
+        if self.shared_experts is not None:
```

Comment:
same comment, is this used by the released model or not?

Reply:
not adressed

Reply:
In that case the arch should be different! Use normal MLP for mlp and expert for expert! 🤗 
You can set the first layer then just do += we want to avoid codepathes as much as possible